effect/jvm/src/main/scala/scalaz/effect/SafeApp.scala

```diff
@@ -19,16 +26,32 @@ package scalaz.effect
  * }}}
  */
 trait SafeApp extends RTS {
-  type Error
+
+  sealed trait ExitStatus
+  object ExitStatus {
+    case class ExitNow(code: Int) extends ExitStatus
+    case class ExitWhenDone(code: Int, timeout: Duration) extends ExitStatus
+    case object DoNotExit extends ExitStatus
+  }
 
   /**
    * The main function of the application, which will be passed the command-line
-   * arguments to the program.
+   * arguments to the program and has to return an `IO` with the errors fully handled.
    */
-  def run(args: List[String]): IO[Error, Unit]
+  def run(args: List[String]): IO[Void, ExitStatus]
 
   /**
    * The Scala main function, intended to be called only by the Scala runtime.
    */
-  final def main(args0: Array[String]): Unit = unsafePerformIO(run(args0.toList))
+  final def main(args0: Array[String]): Unit =
+    unsafePerformIO(run(args0.toList)) match {
+      case ExitStatus.ExitNow(code) =>
+        sys.exit(code)
+      case ExitStatus.ExitWhenDone(code, timeout) =>
+        unsafeShutdownAndWait(timeout)
+        sys.exit(code)
+      case ExitStatus.DoNotExit =>
+        unsafePerformIO(IO.never.toUnit)
```

Comment:
But what daemon threads do we have? The thread pool is creating non-daemon threads, that's why with the previous implementation the program exited after performing the IO

Reply:
I checked, and without that line it just terminates. I found that to be the most elegant way to keep it running. Do you have any other way in mind?

Reply:
I think the meaning of `DoNotExit` should be: it will not call `System.exit(_)`, not that it will fire up other threads or sleep forever. Maybe it could be changed to `ExitNormally` or something similar. After all, if the user wants this behavior, they can achieve it by sleeping forever or sequencing `never` (which will result in a permanently blocked and wasted thread).

Reply:
Yeah yeah, totally on the same page now, it got it wrong at first. But check my latest changes, I removed the line.